ms2deepscore/spectrum_pair_selection.py

```diff
@@ -0,0 +1,278 @@
+        Set to False if pairs with two equal compounds/fingerprints should be excluded.
+    fix_global_bias
+        Default is True in which case the function aims to get the same amount of pairs for
+        each bin globally. This means it add more than max_pairs_par_bin for some bins and/or
+        some compounds to compensate for lack of such scores in other compounds.
+    random_seed
+        Set to integer if the randomness of the pair selection should be reproducible.
+
+    Returns
+    -------
+    scores
+        Sparse array (List of lists) with cherrypicked scores.
+    """
+    # pylint: disable=too-many-arguments
+    size = fingerprints.shape[0]
+    if random_seed is not None:
+        np.random.seed(random_seed)
+    data, i, j = compute_jaccard_similarity_matrix_cherrypicking(
+        fingerprints,
+        selection_bins,
+        max_pairs_per_bin,
+        include_diagonal,
+        fix_global_bias,
+    )
+    return coo_array((np.array(data), (np.array(i), np.array(j))),
+                      shape=(size, size))
+
+
+@numba.njit
+def compute_jaccard_similarity_matrix_cherrypicking(
+    fingerprints: np.ndarray,
+    selection_bins: np.ndarray = np.array([(x/10, x/10 + 0.1) for x in range(0, 10)]),
+    max_pairs_per_bin: int = 20,
+    include_diagonal: bool = True,
+    fix_global_bias: bool = True,
+) -> np.ndarray:
+    """Returns matrix of jaccard indices between all-vs-all vectors of references
+    and queries.
+
+    Parameters
+    ----------
+    fingerprints
+        Fingerprint vectors as 2D numpy array.
+    selection_bins
+        List of tuples with upper and lower bound for score bins.
+        The goal is to pick equal numbers of pairs for each score bin.
+        Sidenote: bins do not have to be of equal size, nor do they have to cover the entire
+        range of the used scores.
+    max_pairs_per_bin
+        Specifies the desired maximum number of pairs to be added for each score bin.
+    include_diagonal
+        Set to False if pairs with two equal compounds/fingerprints should be excluded.
+    fix_global_bias
+        Default is True in which case the function aims to get the same amount of pairs for
+        each bin globally. This means it add more than max_pairs_par_bin for some bins and/or
+        some compounds to compensate for lack of such scores in other compounds.
+
+    Returns
+    -------
+    scores
+        Sparse array (List of lists) with cherrypicked scores.
+    """
+    # pylint: disable=too-many-locals
+    size = fingerprints.shape[0]
+    scores_data = []
+    scores_i = []
+    scores_j = []
+    # keep track of total bias across bins
+    max_pairs_global = len(selection_bins) * [max_pairs_per_bin]
+    for i in range(size):
+        scores_row = np.zeros(size)
+        for j in range(size):
+            if i == j and not include_diagonal:
+                continue
+            scores_row[j] = jaccard_index(fingerprints[i, :], fingerprints[j, :])
```

Comment:
I was thinking of it. But decided that it is probably not worth it. The scores have to computed once which takes much less time than the MS2DeepScore model training itself. And it would also complicate the code a lot because all those score have to be stored somewhere. Right now the big advantage is that we can simply go through all rows individually.